File: app2/src/main/java/ru/neverlands/anclient/utils/ContactRenderHelper.java
package ru.neverlands.anclient.utils;

import android.graphics.Color;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class ContactRenderHelper {

    public static final class EffectState {
        public final int id;
        public final int count;
        public final String timeout;

        public EffectState(int id, int count, String timeout) {
            this.id = id;
            this.count = count <= 0 ? 1 : count;
            this.timeout = timeout == null ? "" : timeout.trim();
        }
    }

    private ContactRenderHelper() {
    }

    public static String formatNickWithLevel(String nick, int level) {
        String safeNick = nick == null ? "" : nick;
        if (level > 0) {
            return safeNick + " [" + level + "]";
        }
        return safeNick;
    }

    public static int resolveNickColor(int classId) {
        if (classId == 1) {
            return Color.RED;
        }
        if (classId == 2) {
            return Color.parseColor("#008000");
        }
        return Color.WHITE;
    }

    /**
     * Определяет, относится ли контакт к нейтралам (без клана).
     */
    public static boolean isNeutralClanName(String clanName) {
        if (clanName == null) {
            return true;
        }
        String normalized = clanName.trim();
        return normalized.isEmpty() || "none".equalsIgnoreCase(normalized);
    }

    /**
     * Единая цветовая политика:
     * - нейтрал (без клана) всегда белый;
     * - остальное по classId (враг/друг/нейтрал).
     */
    public static int resolveNickColor(int classId, String clanName) {
        if (isNeutralClanName(clanName)) {
            return Color.WHITE;
        }
        return resolveNickColor(classId);
    }

    public static String buildEffectIconUrl(int effectId) {
        return "http://image.neverlands.ru/pinfo/eff_" + effectId + ".gif";
    }

    public static List<EffectState> parseEffectStatesCsv(String effectStatesCsv, String fallbackEffectIdsCsv) {
        LinkedHashMap<Integer, EffectState> byId = new LinkedHashMap<>();
        if (effectStatesCsv != null && !effectStatesCsv.trim().isEmpty()) {
            String[] entries = effectStatesCsv.split(",");
            for (String entry : entries) {
                if (entry == null) {
                    continue;
                }
                String trimmed = entry.trim();
                if (trimmed.isEmpty()) {
                    continue;
                }
                String[] parts = trimmed.split(":", 3);
                int id = parseIntSafe(parts.length > 0 ? parts[0] : "", 0);
                if (id <= 0) {
                    continue;
                }
                int count = parseIntSafe(parts.length > 1 ? parts[1] : "", 1);
                String timeout = parts.length > 2 ? sanitizeEffectStatePart(parts[2]) : "";
                EffectState existing = byId.get(id);
                if (existing == null) {
                    byId.put(id, new EffectState(id, count, timeout));
                } else {
                    int mergedCount = Math.max(1, existing.count) + Math.max(1, count);
                    String mergedTimeout = existing.timeout.isEmpty() ? timeout : existing.timeout;
                    byId.put(id, new EffectState(id, mergedCount, mergedTimeout));
                }
            }
        }
        if (!byId.isEmpty()) {
            return new ArrayList<>(byId.values());
        }
        List<Integer> fallbackIds = parseEffectIdsCsv(fallbackEffectIdsCsv);
        for (Integer id : fallbackIds) {
            if (id != null && id > 0) {
                byId.put(id, new EffectState(id, 1, ""));
            }
        }
        return new ArrayList<>(byId.values());
    }

    public static List<Integer> extractEffectIds(List<EffectState> effectStates) {
        List<Integer> ids = new ArrayList<>();
        if (effectStates == null || effectStates.isEmpty()) {
            return ids;
        }
        Set<Integer> unique = new LinkedHashSet<>();
        for (EffectState state : effectStates) {
            if (state == null || state.id <= 0) {
                continue;
            }
            unique.add(state.id);
        }
        ids.addAll(unique);
        return ids;
    }

    public static String toEffectStatesCsv(List<EffectState> effectStates) {
        if (effectStates == null || effectStates.isEmpty()) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        LinkedHashMap<Integer, EffectState> byId = new LinkedHashMap<>();
        for (EffectState state : effectStates) {
            if (state == null || state.id <= 0) {
                continue;
            }
            EffectState existing = byId.get(state.id);
            if (existing == null) {
                byId.put(state.id, new EffectState(state.id, state.count, state.timeout));
            } else {
                int mergedCount = Math.max(1, existing.count) + Math.max(1, state.count);
                String mergedTimeout = existing.timeout.isEmpty() ? state.timeout : existing.timeout;
                byId.put(state.id, new EffectState(state.id, mergedCount, mergedTimeout));
            }
        }
        for (Map.Entry<Integer, EffectState> entry : byId.entrySet()) {
            EffectState state = entry.getValue();
            if (state == null) {
                continue;
            }
            if (sb.length() > 0) {
                sb.append(',');
            }
            sb.append(state.id)
                    .append(':')
                    .append(Math.max(1, state.count))
                    .append(':')
                    .append(sanitizeEffectStatePart(state.timeout));
        }
        return sb.toString();
    }

    public static String formatEffectCounter(EffectState state) {
        return formatEffectCounterText(state);
    }

    public static String formatEffectCounterText(EffectState state) {
        if (state == null) {
            return "";
        }
        String timeout = normalizeTimeoutToHourMinute(state.timeout);
        return "[x" + Math.max(1, state.count) + "]\n(" + timeout + ")";
    }

    public static String formatEffectCounterHtml(EffectState state) {
        if (state == null) {
            return "";
        }
        String timeout = normalizeTimeoutToHourMinute(state.timeout);
        return "<span style=\"display:inline-block;font-size:75%;line-height:1.05;vertical-align:middle;\">[x"
                + Math.max(1, state.count)
                + "]<br>("
                + timeout
                + ")</span>";
    }

    public static List<Integer> parseEffectIdsCsv(String effectIdsCsv) {
        if (effectIdsCsv == null || effectIdsCsv.trim().isEmpty()) {
            return new ArrayList<>();
        }
        String[] parts = effectIdsCsv.split(",");
        Set<Integer> unique = new LinkedHashSet<>();
        for (String part : parts) {
            if (part == null) {
                continue;
            }
            String trimmed = part.trim();
            if (trimmed.isEmpty()) {
                continue;
            }
            try {
                int value = Integer.parseInt(trimmed);
                if (value > 0) {
                    unique.add(value);
                }
            } catch (Exception ignored) {
            }
        }
        return new ArrayList<>(unique);
    }

    public static String toEffectIdsCsv(List<Integer> effectIds) {
        if (effectIds == null || effectIds.isEmpty()) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        Set<Integer> unique = new LinkedHashSet<>(effectIds);
        for (Integer effectId : unique) {
            if (effectId == null || effectId <= 0) {
                continue;
            }
            if (sb.length() > 0) {
                sb.append(',');
            }
            sb.append(effectId);
        }
        return sb.toString();
    }

    private static int parseIntSafe(String value, int fallback) {
        if (value == null) {
            return fallback;
        }
        try {
            return Integer.parseInt(value.trim());
        } catch (Exception ignored) {
            return fallback;
        }
    }

    private static String sanitizeEffectStatePart(String value) {
        if (value == null) {
            return "";
        }
        return value.trim().replace(",", "").replace("\n", "").replace("\r", "");
    }

    private static String normalizeTimeoutToHourMinute(String timeout) {
        String safe = timeout == null ? "" : timeout.trim();
        if (safe.isEmpty()) {
            return "--:--";
        }
        String[] parts = safe.split(":");
        if (parts.length >= 2) {
            return padTimePart(parts[0]) + ":" + padTimePart(parts[1]);
        }
        return safe;
    }

    private static String padTimePart(String value) {
        String safe = value == null ? "" : value.trim();
        if (safe.length() == 1) {
            return "0" + safe;
        }
        return safe;
    }
}

File: ANClient/ContactRenderHelper.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace ANClient
{
    internal static class ContactRenderHelper
    {
        internal sealed class EffectState
        {
            internal readonly int Id;
            internal readonly int Count;
            internal readonly string Timeout;

            internal EffectState(int id, int count, string timeout)
            {
                Id = id;
                Count = count <= 0 ? 1 : count;
                Timeout = timeout == null ? string.Empty : timeout.Trim();
            }
        }

        internal static bool IsNeutralClanName(string clanName)
        {
            if (string.IsNullOrEmpty(clanName))
                return true;

            return clanName.Trim().Equals("none", StringComparison.OrdinalIgnoreCase);
        }

        internal static string GetInclinationName(int inclination)
        {
            switch (inclination)
            {
                case 4:
                    return "Chaos";
                case 3:
                    return "Sumers";
                case 2:
                    return "Lights";
                case 1:
                    return "Darks";
                default:
                    return "0";
            }
        }

        internal static string GetInclinationSign(string inclination)
        {
            switch (inclination)
            {
                case "1":
                    return "darks.gif";
                case "2":
                    return "lights.gif";
                case "3":
                    return "sumers.gif";
                case "4":
                    return "chaoss.gif";
                case "5":
                    return "light.gif";
                case "6":
                    return "dark.gif";
                case "7":
                    return "sumer.gif";
                case "8":
                    return "chaos.gif";
                case "9":
                    return "angel.gif";
                default:
                    return string.Empty;
            }
        }

        internal static List<EffectState> ParseEffectStatesCsv(string effectStatesCsv, string fallbackEffectIdsCsv)
        {
            var byId = new Dictionary<int, EffectState>();
            var order = new List<int>();
            if (!string.IsNullOrEmpty(effectStatesCsv))
            {
                var entries = effectStatesCsv.Split(',');
                foreach (var entry in entries)
                {
                    if (string.IsNullOrEmpty(entry))
                        continue;

                    var parts = entry.Trim().Split(new[] { ':' }, 3);
                    var id = ParseIntSafe(parts.Length > 0 ? parts[0] : string.Empty, 0);
                    if (id <= 0)
                        continue;

                    var count = ParseIntSafe(parts.Length > 1 ? parts[1] : string.Empty, 1);
                    var timeout = parts.Length > 2 ? SanitizeEffectStatePart(parts[2]) : string.Empty;
                    AddOrMergeEffectState(byId, order, new EffectState(id, count, timeout));
                }
            }

            if (order.Count == 0)
            {
                var fallbackIds = ParseEffectIdsCsv(fallbackEffectIdsCsv);
                foreach (var id in fallbackIds)
                {
                    AddOrMergeEffectState(byId, order, new EffectState(id, 1, string.Empty));
                }
            }

            var result = new List<EffectState>();
            foreach (var id in order)
            {
                result.Add(byId[id]);
            }

            return result;
        }

        internal static List<int> ExtractEffectIds(List<EffectState> effectStates)
        {
            var result = new List<int>();
            if (effectStates == null)
                return result;

            foreach (var state in effectStates)
            {
                if (state == null || state.Id <= 0 || result.Contains(state.Id))
                    continue;

                result.Add(state.Id);
            }

            return result;
        }

        internal static string ToEffectStatesCsv(List<EffectState> effectStates)
        {
            if (effectStates == null || effectStates.Count == 0)
                return string.Empty;

            var byId = new Dictionary<int, EffectState>();
            var order = new List<int>();
            foreach (var state in effectStates)
            {
                AddOrMergeEffectState(byId, order, state);
            }

            var parts = new List<string>();
            foreach (var id in order)
            {
                var state = byId[id];
                parts.Add(state.Id.ToString(CultureInfo.InvariantCulture) + ":" +
                          Math.Max(1, state.Count).ToString(CultureInfo.InvariantCulture) + ":" +
                          SanitizeEffectStatePart(state.Timeout));
            }

            return string.Join(",", parts.ToArray());
        }

        internal static List<int> ParseEffectIdsCsv(string effectIdsCsv)
        {
            var result = new List<int>();
            if (string.IsNullOrEmpty(effectIdsCsv))
                return result;

            var parts = effectIdsCsv.Split(',');
            foreach (var part in parts)
            {
                var id = ParseIntSafe(part, 0);
                if (id > 0 && !result.Contains(id))
                {
                    result.Add(id);
                }
            }

            return result;
        }

        internal static string ToEffectIdsCsv(List<int> effectIds)
        {
            if (effectIds == null || effectIds.Count == 0)
                return string.Empty;

            var result = new List<string>();
            foreach (var effectId in effectIds)
            {
                if (effectId > 0 && !result.Contains(effectId.ToString(CultureInfo.InvariantCulture)))
                {
                    result.Add(effectId.ToString(CultureInfo.InvariantCulture));
                }
            }

            return string.Join(",", result.ToArray());
        }

        internal static string FormatEffectCounterText(EffectState state)
        {
            if (state == null)
                return string.Empty;

            return "[x" + Math.Max(1, state.Count).ToString(CultureInfo.InvariantCulture) + "] (" + NormalizeTimeoutToHourMinute(state.Timeout) + ")";
        }

        internal static string FormatEffectCounterCompactText(EffectState state)
        {
            if (state == null)
                return string.Empty;

            return "[x" + Math.Max(1, state.Count).ToString(CultureInfo.InvariantCulture) + "](" + NormalizeTimeoutToHourMinute(state.Timeout) + ")";
        }

        internal static string FormatEffectCounterHtml(EffectState state)
        {
            if (state == null)
                return string.Empty;

            return "<span style=\"display:inline-block;font-size:75%;line-height:1.05;vertical-align:middle;\">[x" +
                   Math.Max(1, state.Count).ToString(CultureInfo.InvariantCulture) +
                   "]<br>(" +
                   NormalizeTimeoutToHourMinute(state.Timeout) +
                   ")</span>";
        }

        internal static string BuildEffectIconsHtml(string effectStatesCsv, string fallbackEffectIdsCsv)
        {
            var states = ParseEffectStatesCsv(effectStatesCsv, fallbackEffectIdsCsv);
            if (states.Count == 0)
                return string.Empty;

            var sb = new StringBuilder();
            foreach (var state in states)
            {
                if (state == null || state.Id <= 0)
                    continue;

                sb.Append("<span class=\"an-room-effect-wrap\">");
                sb.Append("<img class=\"an-room-injury-icon\" src=http://image.neverlands.ru/pinfo/eff_");
                sb.Append(state.Id.ToString(CultureInfo.InvariantCulture));
                sb.Append(".gif border=0 width=15 height=12 title=\"эффект #");
                sb.Append(state.Id.ToString(CultureInfo.InvariantCulture));
                sb.Append("\" align=absmiddle>");
                sb.Append(FormatEffectCounterHtml(state));
                sb.Append("</span>");
            }

            return sb.ToString();
        }

        private static void AddOrMergeEffectState(Dictionary<int, EffectState> byId, List<int> order, EffectState state)
        {
            if (state == null || state.Id <= 0)
                return;

            EffectState existing;
            if (!byId.TryGetValue(state.Id, out existing))
            {
                byId.Add(state.Id, state);
                order.Add(state.Id);
                return;
            }

            var mergedCount = Math.Max(1, existing.Count) + Math.Max(1, state.Count);
            var mergedTimeout = string.IsNullOrEmpty(existing.Timeout) ? state.Timeout : existing.Timeout;
            byId[state.Id] = new EffectState(state.Id, mergedCount, mergedTimeout);
        }

        private static int ParseIntSafe(string value, int fallback)
        {
            if (string.IsNullOrEmpty(value))
                return fallback;

            int parsed;
            return int.TryParse(value.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out parsed) ? parsed : fallback;
        }

        private static string SanitizeEffectStatePart(string value)
        {
            if (string.IsNullOrEmpty(value))
                return string.Empty;

            return value.Trim().Replace(",", string.Empty).Replace("\n", string.Empty).Replace("\r", string.Empty);
        }

        private static string NormalizeTimeoutToHourMinute(string timeout)
        {
            var safe = timeout == null ? string.Empty : timeout.Trim();
            if (safe.Length == 0)
                return "--:--";

            var parts = safe.Split(':');
            if (parts.Length >= 2)
                return PadTimePart(parts[0]) + ":" + PadTimePart(parts[1]);

            return safe;
        }

        private static string PadTimePart(string value)
        {
            var safe = value == null ? string.Empty : value.Trim();
            return safe.Length == 1 ? "0" + safe : safe;
        }
    }
}
